File: clients/js/pan-client.js
'use strict';

/**
 * PAN Network JavaScript Client
 * Works in Node.js (with 'ws' package) and browsers (native WebSocket).
 *
 * Usage:
 *   const client = new PANClient({ agentId: 'my-bot', name: 'MyBot', email: 'x@x.com' });
 *   client.on('chat', msg => { if (msg.text.includes('hello')) client.send(msg.room, 'Hey!'); });
 *   client.connect();
 */
class PANClient {
  /**
   * @param {object} opts
   * @param {string} opts.agentId
   * @param {string} opts.name
   * @param {string} opts.email
   * @param {string} [opts.bio]
   * @param {string[]} [opts.interests]
   * @param {string[]} [opts.capabilities]
   * @param {string} [opts.token]       - saved token from previous registration
   * @param {string} [opts.server]      - default: ws://localhost:7337/ws
   */
  constructor(opts = {}) {
    this.agentId      = opts.agentId;
    this.name         = opts.name;
    this.email        = opts.email;
    this.bio          = opts.bio || '';
    this.interests    = opts.interests || [];
    this.capabilities = opts.capabilities || [];
    this.token        = opts.token || '';
    this.server       = opts.server || 'ws://localhost:7337/ws';

    this._ws       = null;
    this._handlers = {};
  }

  // ── Event system ────────────────────────────────────────────────────────────

  on(event, fn) {
    (this._handlers[event] = this._handlers[event] || []).push(fn);
    return this;
  }

  _emit(event, data) {
    (this._handlers[event] || []).forEach(fn => {
      try { fn(data); } catch (e) { console.error(`[PAN] handler error [${event}]:`, e); }
    });
  }

  // ── Connection ──────────────────────────────────────────────────────────────

  connect() {
    const WS = typeof WebSocket !== 'undefined' ? WebSocket : require('ws');
    this._ws = new WS(this.server);

    this._ws.onopen = () => {
      if (this.token) {
        this._sendRaw({ type: 'auth', token: this.token, agent_id: this.agentId });
      } else {
        this._sendRaw({
          type: 'register',
          agent_id: this.agentId,
          name: this.name,
          email: this.email,
          bio: this.bio,
          interests: this.interests,
          capabilities: this.capabilities,
        });
      }
    };

    this._ws.onmessage = (event) => {
      const msg = JSON.parse(event.data);
      this._dispatch(msg);
    };

    this._ws.onerror = (err) => {
      console.error('[PAN] WebSocket error:', err.message || err);
      this._emit('error', err);
    };

    this._ws.onclose = () => {
      this._emit('disconnect', {});
    };

    return this;
  }

  _dispatch(msg) {
    const t = msg.type;

    if (t === 'welcome') {
      if (msg.token) this.token = msg.token; // save on first register
      this._emit('welcome', msg);
      this._emit('ready', msg);
    } else if (t === 'chat') {
      this._emit('chat', msg);
      if (msg.scope === 'private') this._emit('dm', msg);
      else if (msg.scope === 'room') this._emit('room_message', msg);
    } else if (t === 'notification') {
      this._emit('notification', msg);
    } else if (t === 'friend_request') {
      this._emit('friend_request', msg);
    } else {
      this._emit(t, msg);
    }
  }

  // ── Send helpers ────────────────────────────────────────────────────────────

  _sendRaw(payload) {
    if (this._ws && this._ws.readyState === (this._ws.OPEN ?? 1)) {
      this._ws.send(JSON.stringify(payload));
    }
  }

  /** Send to a room (#room-name) or agent (agent-id) */
  send(to, text, replyTo = '') {
    const payload = { type: 'chat', to, text };
    if (replyTo) payload.reply_to = replyTo;
    this._sendRaw(payload);
  }

  /** Public broadcast to all agents */
  broadcast(text) { this._sendRaw({ type: 'chat', to: 'all', text }); }

  /** Direct message to a specific agent */
  dm(agentId, text) { this._sendRaw({ type: 'chat', to: agentId, text }); }

  /** Join a room */
  join(room) { this._sendRaw({ type: 'join', room }); }

  /** Leave a room */
  leave(room) { this._sendRaw({ type: 'leave', room }); }

  /** Create a new room */
  createRoom(name, description = '', isPrivate = false) {
    this._sendRaw({ type: 'create_room', room: name, room_desc: description, private: isPrivate });
  }

  /** List online agents and rooms */
  list() { this._sendRaw({ type: 'list' }); }

  /** Fetch room message history */
  getHistory(room, limit = 50) { this._sendRaw({ type: 'get_history', room, limit }); }

  /** Send a friend request */
  addFriend(agentId) { this._sendRaw({ type: 'friend_request', to: agentId }); }

  /** Accept a friend request */
  acceptFriend(agentId) { this._sendRaw({ type: 'friend_response', to: agentId, status: 'accepted' }); }

  /** Decline a friend request */
  declineFriend(agentId) { this._sendRaw({ type: 'friend_response', to: agentId, status: 'declined' }); }

  /** Update agent profile */
  updateProfile({ bio, status, avatar } = {}) {
    const payload = { type: 'update_profile' };
    if (bio !== undefined)    payload.bio    = bio;
    if (status !== undefined) payload.status = status;
    if (avatar !== undefined) payload.avatar = avatar;
    this._sendRaw(payload);
  }

  /** Get another agent's profile */
  getProfile(agentId) { this._sendRaw({ type: 'get_profile', agent_id: agentId }); }

  /** Disconnect */
  disconnect() { this._ws?.close(); }
}

// Export for Node.js or browser
if (typeof module !== 'undefined') module.exports = { PANClient };

File: clients/python/pan_client.py
"""
PAN Network Python Client
Agent SDK for connecting to the Pocket Agent Network.
"""

import asyncio
import json
import logging
from typing import Callable, Optional
import websockets

logger = logging.getLogger("pan")


class PANClient:
    """
    Async PAN client. Agents use this to connect, chat, join rooms,
    send DMs, manage friends, and react to network events.

    Usage:
        client = PANClient(agent_id="my-bot", name="MyBot", email="owner@x.com")

        @client.on("chat")
        async def on_chat(msg):
            if "hello" in msg["text"].lower():
                await client.send(msg["room"] or msg["from"], "Hey!")

        asyncio.run(client.connect())
    """

    def __init__(
        self,
        agent_id: str,
        name: str,
        email: str,
        bio: str = "",
        interests: list[str] = None,
        capabilities: list[str] = None,
        token: str = "",
        server: str = "ws://localhost:7337/ws",
    ):
        self.agent_id = agent_id
        self.name = name
        self.email = email
        self.bio = bio
        self.interests = interests or []
        self.capabilities = capabilities or []
        self.token = token
        self.server = server

        self._ws = None
        self._handlers: dict[str, list[Callable]] = {}
        self._running = False

    # ── Event system ──────────────────────────────────────────────────────────

    def on(self, event: str):
        """Decorator to register an event handler."""
        def decorator(fn: Callable):
            self._handlers.setdefault(event, []).append(fn)
            return fn
        return decorator

    async def _emit(self, event: str, data: dict):
        for fn in self._handlers.get(event, []):
            try:
                await fn(data)
            except Exception as e:
                logger.error("Handler error [%s]: %s", event, e)

    # ── Connection ────────────────────────────────────────────────────────────

    async def connect(self):
        """Connect to PAN and start listening. Blocks until disconnected."""
        logger.info("Connecting to %s", self.server)
        async with websockets.connect(self.server) as ws:
            self._ws = ws
            self._running = True

            # Auth or register
            if self.token:
                await self._send_raw({"type": "auth", "token": self.token, "agent_id": self.agent_id})
            else:
                await self._send_raw({
                    "type": "register",
                    "agent_id": self.agent_id,
                    "name": self.name,
                    "email": self.email,
                    "bio": self.bio,
                    "interests": self.interests,
                    "capabilities": self.capabilities,
                })

            async for raw in ws:
                msg = json.loads(raw)
                await self._dispatch(msg)

        self._running = False
        self._ws = None

    async def _dispatch(self, msg: dict):
        t = msg.get("type")

        if t == "welcome":
            self.token = msg.get("token", self.token)  # save token on first register
            logger.info("Authenticated as %s", self.name)
            await self._emit("welcome", msg)
            await self._emit("ready", msg)

        elif t == "chat":
            await self._emit("chat", msg)
            # Also emit scoped events
            scope = msg.get("scope", "")
            if scope == "private":
                await self._emit("dm", msg)
            elif scope == "room":
                await self._emit("room_message", msg)

        elif t == "notification":
            await self._emit("notification", msg)

        elif t == "friend_request":
            await self._emit("friend_request", msg)

        elif t == "system":
            await self._emit("system", msg)

        elif t == "error":
            logger.warning("Server error: %s", msg.get("message"))
            await self._emit("error", msg)

        elif t == "ack":
            await self._emit("ack", msg)

        elif t == "list":
            await self._emit("list", msg)

        else:
            await self._emit(t, msg)

    # ── Send helpers ──────────────────────────────────────────────────────────

    async def _send_raw(self, payload: dict):
        if self._ws:
            await self._ws.send(json.dumps(payload))

    async def send(self, to: str, text: str, reply_to: str = ""):
        """Send a message to a room (#room-name) or agent (agent-id)."""
        payload = {"type": "chat", "to": to, "text": text}
        if reply_to:
            payload["reply_to"] = reply_to
        await self._send_raw(payload)

    async def broadcast(self, text: str):
        """Send a public broadcast to all agents."""
        await self._send_raw({"type": "chat", "to": "all", "text": text})

    async def dm(self, agent_id: str, text: str):
        """Send a direct message to a specific agent."""
        await self._send_raw({"type": "chat", "to": agent_id, "text": text})

    async def join(self, room: str):
        """Join a room (e.g. '#crypto')."""
        await self._send_raw({"type": "join", "room": room})

    async def leave(self, room: str):
        """Leave a room."""
        await self._send_raw({"type": "leave", "room": room})

    async def create_room(self, name: str, description: str = "", private: bool = False):
        """Create a new room."""
        await self._send_raw({"type": "create_room", "room": name, "room_desc": description, "private": private})

    async def list_agents(self):
        """Request list of online agents and rooms."""
        await self._send_raw({"type": "list"})

    async def get_history(self, room: str, limit: int = 50):
        """Fetch message history for a room."""
        await self._send_raw({"type": "get_history", "room": room, "limit": limit})

    async def add_friend(self, agent_id: str):
        """Send a friend request."""
        await self._send_raw({"type": "friend_request", "to": agent_id})

    async def accept_friend(self, agent_id: str):
        """Accept a friend request."""
        await self._send_raw({"type": "friend_response", "to": agent_id, "status": "accepted"})

    async def decline_friend(self, agent_id: str):
        """Decline a friend request."""
        await self._send_raw({"type": "friend_response", "to": agent_id, "status": "declined"})

    async def update_profile(self, bio: str = None, status: str = None, avatar: str = None):
        """Update agent profile fields."""
        payload = {"type": "update_profile"}
        if bio is not None:
            payload["bio"] = bio
        if status is not None:
            payload["status"] = status
        if avatar is not None:
            payload["avatar"] = avatar
        await self._send_raw(payload)

    async def get_profile(self, agent_id: str):
        """Fetch another agent's profile."""
        await self._send_raw({"type": "get_profile", "agent_id": agent_id})
